### PR TITLE
[fix-10608] [worker] move setting of TaskExecutionContext to finally block

### LINKED ISSUE #10608: [Bug] [Worker] TaskInstance state update error
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

find version: 2.0.1

When TaskExecuteThread  is running, if an exception is threw, it is caught and the kill method is executed,  but if kill methond also throw error instead of an exception, the `TaskExecuteResponseCommand` can not be set properties, like status, endTime, these can be set default value, for example, the `status` can be set 0 (state of `submit success`)

- log
image

- code
image

### What you expected to happen

The state of TaskInstance is correct

### How to reproduce

- 1. throw `exception/error `in try block
- 2. throw `error `in kill methond

### Anything else

_No response_

### Version

2.0.3

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by guoshupei**
## Purpose of the pull request
fix #10608 

## Brief change log
- move setting to finally
- add log

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

### CODE REVIEW

File: dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java [hunk 1]
```diff
@@ -191,21 +192,20 @@ public void run() {
...
+        } finally {
+            taskExecutionContext.setCurrentExecutionStatus(errorFlag ? ExecutionStatus.FAILURE : ExecutionStatus.of(this.task.getExitStatus().getCode()));
             taskExecutionContext.setEndTime(DateUtils.getCurrentDate());
             taskExecutionContext.setProcessId(this.task.getProcessId());
             taskExecutionContext.setAppIds(this.task.getAppIds());
-        } finally {
+            if (!errorFlag) {
+                taskExecutionContext.setVarPool(JSONUtils.toJsonString(this.task.getParameters().getVarPool()));
+            }
+
```
Comment (ruanwenjun) [inline dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java]:
```suggestion
        } finally {
                    if(errorFlag) {
                        taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.FAILURE);
                    }else {
taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.of(this.task.getExitStatus().getCode()));    
                    }
                    taskExecutionContext.setEndTime(DateUtils.getCurrentDate());
                    taskExecutionContext.setProcessId(this.task.getProcessId());
                    taskExecutionContext.setAppIds(this.task.getAppIds());
```

**Review by ruanwenjun — COMMENTED** [review]
And please revert the change of setVarPool.
```
taskExecutionContext.setVarPool(JSONUtils.toJsonString(this.task.getParameters().getVarPool()));
```

**Review by guoshupei — COMMENTED** [review]
change and  commit again

**Comment by guoshupei** [comment]
> And please revert the change of setVarPool.
> 
> ```
> taskExecutionContext.setVarPool(JSONUtils.toJsonString(this.task.getParameters().getVarPool()));
> ```

good idea.

**Review by ruanwenjun — APPROVED** [review]
LGTM

File: dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java [hunk 2]
```diff
@@ -192,20 +193,23 @@ public void run() {
...
             kill();
-            taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.FAILURE);
+        } finally {
+            if (errorFlag) {
+                taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.FAILURE);
+            } else {
+                taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.of(this.task.getExitStatus().getCode()));
+            }
             taskExecutionContext.setEndTime(DateUtils.getCurrentDate());
             taskExecutionContext.setProcessId(this.task.getProcessId());
```
Comment (caishunfeng) [inline dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:210]: If in finally, the task maybe null, so it need to add NPE check.

Comment (guoshupei) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:210]: > If in finally, the task maybe null, so it need to add NPE check.

yes, I check again. you are right. if add NPE check, I think move setting in try block and catch block is better. it readability is better. WDYT?

Comment (ruanwenjun) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:210]: It seems you didn't handle the task may be null in finally.

Comment (ruanwenjun) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:210]: You can use executionStatus to replace the errorFlag, then you don't need to deal with the task in finally.

Comment (guoshupei) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:210]: How do I determine whether a task is null through executionStatus？There may be the same value` ExecutionStatus.FAILURE` in try block and catch block and task may also be null.  I think if task throw NPE, it need to add NPE check in both catch block and finally block . so I think add check in finally is better.

Comment (ruanwenjun) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:210]: You can define a ExecutionStatus and set the value at catch, just like errorFlag, then you don't need to parse the status from task.
```java
ExecutionStatus executionStatus;
try{
executionStatus = ExecutionStatus.of(this.task.getExitStatus().getCode());
}catch(Exception) {
executionStatus = ExecutionStatus.FAILURE;
}finally{
taskExecutionContext.setCurrentExecutionStatus(executionStatus);
}
```

Comment (guoshupei) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:210]: the task is also used in getProcessId() and getAppIds() , so it need to add NPE check

**Comment by caishunfeng** [comment]
No update for months, the lastest code had changed, so I will close this pr. Feel free to reopen it if need.

**Pull request closed without merging**